src/render/ext/device.rs

```diff
@@ -71,42 +74,30 @@ impl<D: device::Device> Graphics<D> {
         self.device.submit(self.renderer.as_buffer());
         self.renderer.reset();
     }
+    
+    /// Cleanup resources after the frame.
+    pub fn cleanup(&mut self) {
+        self.device.after_frame();
+        self.factory.cleanup();
+    }
 }
 
 
 /// Backend extension trait for convenience methods
-pub trait DeviceExt<R: device::Resources, C: device::draw::CommandBuffer<R>>:
-    device::Factory<R> + device::Device<Resources = R, CommandBuffer = C>
-{
-    /// Create a new renderer
-    fn create_renderer(&mut self) -> ::Renderer<R, C>;
+pub trait DeviceExt<D: device::Device, F> {
     /// Convert to single-threaded wrapper
-    fn into_graphics(mut self) -> Graphics<Self>;
+    fn into_graphics(mut self) -> Graphics<D, F>;
```

Comment:
Does it matter? I don't know the guidelines w.r.t associated types, but it's not obvious to me that they'd improve this particular code in any way.